### PR TITLE
Add Solar Eclipses to Astronomical calculations

### PR DESCRIPTION
**Pull request by worachate001**
### Description
This PR will fix issue #407 by adding the method to search for all solar eclipses during user-input timespan.

The calculations are divided into two tabs.

First tab for generic information of all solar eclipses:
* Date and time of greatest eclipse
* Saros
* Type of eclipses
* Gamma
* Magnitude at greatest eclipse
* Geographic coordinates of greatest eclipse
* Altitude of the Sun at greatest eclipse
* Path Width of umbra/antumbra
* Central duration of totality/annularity
* Double click the event will go to the location and time of greatest eclipse (need help to handle it properly when location changed -- **already fixed**, thanks to @alex-w)

Second tab for information of solar eclipses visible at current location
* Date of eclipse
* Time of each contacts and maximum eclipse
* Magnitude of maximum eclipse
* Duration of totality/annularity if the eclipse is total/annular
* Double click the event will go to the time of maximum eclipse

Notes:
* I'm still testing the tool with many situations to search for possible bug.
* Lunar eclipse calculations are also improved by removing unnecessary steps.

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] This change requires a documentation update

### How Has This Been Tested?

**Test Configuration**:
* Operating system: Ubuntu 20.04
* Graphics Card: NVidia Geforce RTX

## Checklist:

- [x] My code follows the [code style](http://stellarium.org/doc/head/codingStyle.html) of this project.
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (header file)
- [ ] I have updated the respective chapter in the Stellarium User Guide
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Review by alex-w — CHANGES_REQUESTED** [review]
Thanks for adding new features!

I see problem: in the tab "All Solar Eclipses" after double click the info about location is not changes - this is wrong!

**Comment by worachate001** [comment]
> Thanks for adding new features!
> 
> I see problem: in the tab "All Solar Eclipses" after double click the info about location is not changes - this is wrong!

Yes, I need your help about that. The location changed, but not the info.

**Comment by alex-w** [comment]
> Yes, I need your help about that. The location changed, but not the info.

In _AstroCalcDialog::selectCurrentSolarEclipse_:

```
StelLocation loc = core->getCurrentLocation();
loc.latitude = lat;
loc.longitude = lon;
loc.altitude = 10; // 10 meters above sea level
loc.name = ""; // Or some good name
core->moveObserverTo(loc, 0.);

if (objectMgr->findAndSelect(name)) // "name" has English only name
```

**Comment by alex-w** [comment]
I guess the first column (Date and Time) should be in UTC for tab "All Solar Eclipses". @gzotti?

**Comment by gzotti** [comment]
I am in a workshop and cannot look at it closely. My natural feeling is it should be user time zone. But I don't see the GUI context now. Can answer in 6 hours.

File: src/gui/AstroCalcDialog.cpp [hunk 1]
```diff
@@ -3132,7 +3132,7 @@ void AstroCalcDialog::generateSolarEclipses()
 					int nx = -61 * nd;
 					int nc = floor(nx / 358. + 0.5 - nd / (12. * 358 * 358));
 					int saros = 1 + ((s + nc * 223 - 1) % 223);
-					if ((s + nc * 223 - 1) < 0) saros -= 223;
+					if ((s + nc * 223 - 1) < 0)saros -= 223;
```
Comment (alex-w) [inline src/gui/AstroCalcDialog.cpp:3135]: Missing space?

**Comment by alex-w** [comment]
> I am in a workshop and cannot look at it closely. My natural feeling is it should be user time zone. But I don't see the GUI context now. Can answer in 6 hours.

So, it shgould be as is - thanks!

File: src/gui/AstroCalcDialog.cpp (hunk 1 above)
Comment (worachate001) [reply src/gui/AstroCalcDialog.cpp:3135]: Thank you, fix on the way.

**Comment by alex-w** [comment]
I have question for “Altitude” column - probably this data calculated due it presented on NASA eclipses website - why it rounded to whole?

**Comment by worachate001** [comment]
> I have question for “Altitude” column - probably this data calculated due it presented on NASA eclipses website - why it rounded to whole?

It's altitude with no refraction, so they are not very accurate. Altitude is useful because you will know that the Sun is high or low above the horizon during the eclipse.

**Comment by alex-w** [comment]
I'm afraid that in _selectCurrentSolarEclipse_ there should be something like:

```
StelLocation loc;
loc.latitude = lat;
loc.longitude = lon;
loc.altitude = 10; // 10 meters above sea level
loc.name = q_(“Greatest eclipse’s point”);
loc.planetName = "Earth”;
loc.ianaTimeZone = “LMST”;
core->moveobserver(location, 0.);
```

**Comment by worachate001** [comment]
I will note that in some dates, especially annular eclipse with short duration, you can see that there is a little difference between position of the Moon and Sun at the place of greatest eclipse. I'm still looking for the reason behind it (maybe about the Besselian elements).

**Comment by alex-w** [comment]
Maybe column “Saros Number” should be identical for all subtabs and just “Saros”?

**Comment by worachate001** [comment]
> Maybe column “Saros Number” should be identical for all subtabs and just “Saros”?

I agree, thinking about that too.

**Comment by alex-w** [comment]
Please add translatable tooltips for data in columns

**Comment by worachate001** [comment]
> Please add translatable tooltips for data in columns

Thank you for suggestion, I'm adding some of them

**Comment by alex-w** [comment]
Maybe synchronize dates in the input fields for Solar eclipses will good idea, e.g. if user changes the startup date for tab "All Solar Eclipses", then set it as a startup date in the tab "Local Solar Eclipses" also?

@gzotti?

**Review by alex-w — APPROVED** [review]

**Comment by gzotti** [comment]
Great addition, thank you! However, I really want to also have good documentation of our sources. Presumably you did not invent the way of computing from scratch, right? Did you use some algorithm that is available in the literature? Please follow usual standards, at least author, book title, publisher and year. And ISBN/DOI or URL for a modern source where available.

File: src/gui/AstroCalcDialog.cpp [hunk 2]
```diff
@@ -2587,6 +2597,1227 @@ void AstroCalcDialog::saveLunarEclipses()
...
+									JD = JD3;
+									JD3 = JD2;
+									JD2 = JD;
+								}
+								else
+									eclipseTypeStr = "Annular";
+							}
+							else
+								{
+									eclipseTypeStr = "Partial";
```
Comment (alex-w) [inline src/gui/AstroCalcDialog.cpp:3541]: Please fix translatability

Hint: `qc_("Partial", "eclipse type");`

File: src/gui/AstroCalcDialog.cpp [hunk 3]
```diff
@@ -2587,6 +2597,1227 @@ void AstroCalcDialog::saveLunarEclipses()
...
+								centraleclipse = true;
+								if (L2 < 0.)
+								{
+									eclipseTypeStr = "Total";
+									JD = JD3;
+									JD3 = JD2;
+									JD2 = JD;
+								}
+								else
+									eclipseTypeStr = "Annular";
```
Comment (alex-w) [inline src/gui/AstroCalcDialog.cpp:3537]: Please fix translatability

Hint: `qc_("Annular", "eclipse type");`

File: src/gui/AstroCalcDialog.cpp [hunk 4]
```diff
@@ -2587,6 +2597,1227 @@ void AstroCalcDialog::saveLunarEclipses()
...
+							double C3altitude = eclipseData.altitude;
+
+							L2 = eclipseData.L2;
+
+							if (abs(JD3-JD2) > 1./86400. && (C2altitude > 0.) && (C3altitude > 0.)) // Central eclipse occurs
+							{
+								centraleclipse = true;
+								if (L2 < 0.)
+								{
+									eclipseTypeStr = "Total";
```
Comment (alex-w) [inline src/gui/AstroCalcDialog.cpp:3531]: Please fix translatability

Hint: `qc_("Total", "eclipse type");`

**Review by alex-w — APPROVED** [review]
Thanks!

**Pull request merged**